src/middlewared/middlewared/api/v25_04_0/nfs.py

```diff
@@ -0,0 +1,194 @@
+EXCLUDED_PORTS = [NFS_RDMA_DEFAULT_PORT]
+
+
+class NfsEntry(BaseModel):
+    id: int
+    servers: Annotated[int | None, Field(ge=1, le=256)]
+    """ Specify the number of nfsd. Default: Number of nfsd is equal number of CPU. """
+    allow_nonroot: bool
+    """ Allow non-root mount requests.  This equates to 'insecure' share option. """
+    protocols: list[NFS_protocols]
+    """ Specify supported NFS protocols:  NFSv3, NFSv4 or both can be listed. """
+    v4_krb: bool
+    """ Force Kerberos authentication on NFS shares. """
+    v4_domain: str
+    """ Specify a DNS domain (NFSv4 only) """
+    bindip: list[NonEmptyString] = Field(default_factory=[])  # Using IPvAnyAddress creates JSON serialization errors
+    """ Limit the server IP addresses available for NFS """
+    mountd_port: Annotated[TcpPort | None, AfterValidator(exclude_tcp_ports(EXCLUDED_PORTS))]
+    """ Specify the mountd port binding """
+    rpcstatd_port: Annotated[TcpPort | None, AfterValidator(exclude_tcp_ports(EXCLUDED_PORTS))]
+    """ Specify the rpc.statd port binding """
+    rpclockd_port: Annotated[TcpPort | None, AfterValidator(exclude_tcp_ports(EXCLUDED_PORTS))]
+    """ Specify the rpc.lockd port binding """
+    mountd_log: bool
+    """ Enable or disable mountd logging """
+    statd_lockd_log: bool
+    """ Enable or disable statd and lockd logging """
+    v4_krb_enabled: bool
+    """ Status of NFSv4 authentication requirement (status only) """
+    userd_manage_gids: bool
+    """ Enable to allow server to manage gids """
+    keytab_has_nfs_spn: bool
+    """ Report status of NFS Principal Name in keytab (status only)"""
+    managed_nfsd: bool
+    """ Report status of 'servers' field.
+    If True the number of nfsd are managed by the server. (status only)"""
+    rdma: bool
+    """ Enable or disable NFS over RDMA.  Requires RDMA capable NIC """
+
+    @field_validator('bindip')
+    @classmethod
+    def check_bind_ip(cls, field_value: list):
+        """ Custom validator for IP addresses to avoid JSON serialization errors """
+        all(isinstance(v, IPvAnyAddress) for v in field_value)
+        return field_value
+
+
+@single_argument_args('nfs_update')
+class NfsUpdateArgs(NfsEntry, metaclass=ForUpdateMetaclass):
+    id: Excluded = excluded_field()
+    managed_nfsd: Excluded = excluded_field()
+    v4_krb_enabled: Excluded = excluded_field()
+    keytab_has_nfs_spn: Excluded = excluded_field()
+
+
+class NfsUpdateResult(BaseModel):
+    result: NfsEntry
+
+
+class NfsBindipChoicesArgs(BaseModel):
+    pass
+
+
+@single_argument_result
+class NfsBindipChoicesResult(BaseModel):
+    """ Return a dictionary of IP addresses """
+    model_config = ConfigDict(extra='allow')
+
+
+class NfsShareEntry(BaseModel):
+    id: int
+    path: NonEmptyString
+    """ Local path to be exported. """
+    aliases: list[NonEmptyString] = []
+    """ IGNORED for now """
+    comment: str = ""
+    """ User comment associated with share """
```

Comment:
Most docstrings (for `NfsShareEntry`) appear to be terminated by a period.  Maybe be consistent.  Meh.